core/http/src/uri/uri.rs

```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
+            Uri::try_from(v.to_string()).map_err(de::Error::custom)
+        }
+        fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
+        where
+                E: de::Error, {
+            Uri::try_from(v).map_err(de::Error::custom)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for Uri<'de> {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+                D: Deserializer<'de> {
+            deserializer.deserialize_str(UriVisitor)
+        }
+    }
+    
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        
+        use std::convert::TryFrom;
+        use crate::{ext::IntoOwned, uri::{self, Origin, Authority, Absolute, Host}};
+        #[derive(Debug)]
+        enum TestError {
+            UriError(Box<uri::Error<'static>>),
+            JsonError(serde_json::Error),
+        }
+        impl<'a> From<uri::Error<'a>> for TestError {
+            fn from(e: uri::Error<'a>) -> Self {
+                Self::UriError(Box::new(e.into_owned()))
+            }
+        }
+        impl From<serde_json::Error> for TestError {
+            fn from(e: serde_json::Error) -> Self {
+                Self::JsonError(e)
+            }
+        }
+
+        // Test Helper function
+        // 'a sould almost always be 'static, but it may be useful to not require that
+        fn test_helper<'a>(raw: &str, json: &'a str, uri_obj: Uri<'a>) -> Result<(), TestError> {
+            let uri = Uri::try_from(raw).expect(raw);
+            assert_eq!(uri, uri_obj, "Uri::try_from did not create the expected Uri: {} != {}", raw, uri_obj);
+
+            assert_eq!(serde_json::to_string(&uri)?, json, "Serialization did not produce the expected Json string");
+            assert_eq!(serde_json::from_str::<Uri<'a>>(json)?, uri, "Deserialization from a `&str` did not produce the expected Uri");
+            Ok(())
+        }
+        
+        /// /path?query
+        fn origin<'a>(path: &'a str, query: Option<&'a str>) -> Uri<'a> {
+            Uri::Origin(Origin::new(path, query))
+        }
+
+        #[test]
+        fn serde_origin_form() -> Result<(), TestError> {
+            test_helper("/", "\"/\"", origin("/", None))?;
+            test_helper("/foo", "\"/foo\"", origin("/foo", None))?;
+            test_helper("/bar/foo", "\"/bar/foo\"", origin("/bar/foo", None))?;
+            test_helper("/index.html", "\"/index.html\"", origin("/index.html", None))?;
+            test_helper("/a/nice/url", "\"/a/nice/url\"", origin("/a/nice/url", None))?;
+
+            test_helper("/a?param", "\"/a?param\"", origin("/a", Some("param")))?;
+            test_helper("/a?param=value", "\"/a?param=value\"", origin("/a", Some("param=value")))
+        }
+
+        /// username:password@some.host:8088
+        fn authority<'a>(user: Option<&'a str>, host: &'a str, port: Option<u16>) -> Uri<'a> {
+            Uri::Authority(Authority::new(user, Host::Raw(host), port))
+        }
```

Comment:
Same here as with `origin()`.

Reply:
It actually is nessecary, at least with serde_test. The type passed to `assert_tokens` is the type it serializes to and deserialized from.

Reply:
Why does that matter? All of the URI variants implement `Into<Uri>`. You could simply make `test_helper` generic on a `T: Into<Uri>` and pass in an `Origin` or `Authority` directly.